File: backend/src/main/java/com/hgw/gestionale/product/controller/ProductController.java
package com.hgw.gestionale.product.controller;

import com.hgw.gestionale.product.dto.CreateProductRequest;
import com.hgw.gestionale.product.dto.ProductResponse;
import com.hgw.gestionale.product.dto.UpdateProductRequest;
import com.hgw.gestionale.product.service.ProductService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping(value = "/api/products", produces = MediaType.APPLICATION_JSON_VALUE)
@RequiredArgsConstructor
@Slf4j
public class ProductController {
    private final ProductService productService;

    @GetMapping
    public List<ProductResponse> getAllProducts() {
        log.info("ProductController.getAll richiesta lista product");
        return productService.getAll();
    }

    @GetMapping("/{id}")
    public ProductResponse getProductById(@PathVariable Long id) {
        log.info("ProductController.getById richiesta product id={}", id);
        return productService.getById(id);
    }

    @GetMapping("/by-seriale/{seriale}")
    public ProductResponse getProductBySeriale(@PathVariable String seriale) {
        log.info("ProductController.getBySeriale richiesta product seriale={}", seriale);
        return productService.getBySeriale(seriale);
    }

    @GetMapping("/by-imei/{imei}")
    public ProductResponse getProductByImei(@PathVariable String imei) {
        log.info("ProductController.getByImei richiesta product imei={}", imei);
        return productService.getByImei(imei);
    }

    @GetMapping("/search")
    public List<ProductResponse> searchProduct(@RequestParam String modelNome) {
        log.info("ProductController.search ricerca product modelNome={}", modelNome);
        return productService.search(modelNome);
    }

    @PutMapping("/{id}")
    public ProductResponse updateProduct(
            @PathVariable Long id,
            @Valid @RequestBody UpdateProductRequest request
    ) {
        log.info("ProductController.update aggiornamento product id={}", id);
        return productService.update(id, request);
    }

    @DeleteMapping("/{id}")
    public void deleteProduct(@PathVariable Long id) {
        log.info("ProductController.delete eliminazione product id={}", id);
        productService.delete(id);
    }
}

File: backend-node/src/controller/productController.ts
import {
    Body,
    Controller,
    Delete,
    Get,
    OperationId,
    Patch,
    Path,
    Put,
    Query,
    Route,
    Security,
    Tags,
} from "tsoa";
import { ProductResponse, UpdateProductRequest, UpdateTestDiagnosticiRequest, ProductSortBy } from "../dto/product.dto";
import { PaginatedResponse, SortOrder } from "../common/pagination";
import { ProductService } from "../service/product.service";

const productService = new ProductService();

@Route("products")
@Tags("Products")
@Security("bearerAuth")
export class ProductController extends Controller {
    @Get("/")
    @OperationId("getAllProducts")
    public async getAllProducts(
        @Query() page?: number,
        @Query() pageSize?: number,
        @Query() sortBy?: ProductSortBy,
        @Query() sortOrder?: SortOrder,
        @Query() modello?: string,
        @Query() brand?: string
    ): Promise<PaginatedResponse<ProductResponse>> {
        return productService.getAll(page, pageSize, sortBy, sortOrder, modello, brand);
    }

    @Get("/search")
    @OperationId("searchProducts")
    public async searchProducts(@Query() modelNome: string): Promise<ProductResponse[]> {
        return productService.search(modelNome);
    }

    @Get("/by-seriale/{seriale}")
    @OperationId("getProductBySeriale")
    public async getProductBySeriale(@Path() seriale: string): Promise<ProductResponse> {
        return productService.getBySeriale(seriale);
    }

    @Get("/by-imei/{imei}")
    @OperationId("getProductByImei")
    public async getProductByImei(@Path() imei: string): Promise<ProductResponse> {
        return productService.getByImei(imei);
    }

    @Get("/{id}")
    @OperationId("getProductById")
    public async getProductById(@Path() id: number): Promise<ProductResponse> {
        return productService.getById(id);
    }

    @Put("/{id}")
    @OperationId("updateProduct")
    public async updateProduct(
        @Path() id: number,
        @Body() body: UpdateProductRequest
    ): Promise<ProductResponse> {
        return productService.update(id, body);
    }

    @Patch("/{id}/test-diagnostici")
    @OperationId("updateProductTestDiagnostici")
    public async updateProductTestDiagnostici(
        @Path() id: number,
        @Body() body: UpdateTestDiagnosticiRequest
    ): Promise<ProductResponse> {
        return productService.updateTestDiagnostici(id, body);
    }

    @Delete("/{id}")
    @OperationId("deleteProduct")
    public async deleteProduct(@Path() id: number): Promise<void> {
        return productService.delete(id);
    }
}
